// sppurge/dist/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Delete = void 0;
var sprequest = require("sp-request");
var utils_1 = require("./../utils");
var Delete = (function () {
    function Delete() {
        var _this = this;
        this.getCachedRequest = function (context) {
            return _this.spr || sprequest.create(context.creds);
        };
    }
    Delete.prototype.deleteFile = function (context, filePath) {
        var restUrl = context.siteUrl + "/_api/Web/GetFileByServerRelativeUrl(@FilePath)" +
            ("?@FilePath='" + (0, utils_1.escapeURIComponent)(filePath) + "'");
        return this.deleteRequest(context, restUrl);
    };
    Delete.prototype.deleteFolder = function (context, folderPath) {
        var restUrl = context.siteUrl + "/_api/Web/GetFolderByServerRelativeUrl(@FolderPath)" +
            ("?@FolderPath='" + (0, utils_1.escapeURIComponent)(folderPath) + "'");
        return this.deleteRequest(context, restUrl);
    };
    Delete.prototype.deleteRequest = function (context, restUrl) {
        var _this = this;
        this.spr = this.getCachedRequest(context);
        return this.spr.requestDigest(context.siteUrl).then(function (digest) {
            return _this.spr.post(restUrl, {
                headers: {
                    'X-RequestDigest': digest,
                    'X-HTTP-Method': 'DELETE',
                    'Accept': 'application/json; odata=verbose',
                    'Content-Type': 'application/json; odata=verbose'
                }
            });
        });
    };
    return Delete;
}());
exports.Delete = Delete;
